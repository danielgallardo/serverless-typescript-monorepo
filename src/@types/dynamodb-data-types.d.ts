/* tslint:disable */
declare module 'dynamodb-data-types' {
  export interface AttributeValue {
    B?: string;
    BS?: string[];
    BOOL?: boolean;
    L?: AttributeValue[];
    M?: {[id: string]: AttributeValue};
    N?: string;
    NS?: string[];
    NULL?: boolean;
    S?: string;
    SS?: string[];
  }

  export interface IAttributeValue {
    unwrap(obj: {[key: string]: AttributeValue}): {[key: string]: any};
    wrap(obj: {[key: string]: any}): {[key: string]: AttributeValue};
  }

  export const AttributeValue: IAttributeValue;
}
