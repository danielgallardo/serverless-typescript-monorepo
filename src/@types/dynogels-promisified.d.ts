declare module 'dynogels-promisified' {
  export import AWS = require('aws-sdk');
  import stream = require('stream');
  import {DocumentClient} from 'aws-sdk/lib/dynamodb/document_client';
  import * as joi from 'joi';
  import ConsumedCapacity = DocumentClient.ConsumedCapacity;

  // Dynogels Data Members
  export let log: Log;
  export let models: {[key: string]: Model};
  export let types: {
    stringSet(): joi.AnySchema;
    numberSet(): joi.AnySchema;
    binarySet(): joi.AnySchema;
    uuid(): joi.AnySchema;
    timeUUID(): joi.AnySchema;
  };

  export interface Log {
    info(...args: any[]): void;
    warn(...args: any[]): void;
  }

  // Dynogels global functions
  export function dynamoDriver(dynamoDB: AWS.DynamoDB): AWS.DynamoDB;

  export function reset(): void;

  export function define(modelName: string, config: ModelConfiguration): Model;

  export function createTables(callback: (err: string) => void): void;
  export function createTables(
    options: {[key: string]: CreateTablesOptions} | DynogelsGlobalOptions,
    callback: (err: string) => void
  ): void;

  export function Set(...args: any[]): any;

  export interface DynogelsGlobalOptions {
    $dynogels: {
      pollingInterval: number;
    };
  }

  export interface CreateTablesOptions {
    readCapacity?: number;
    writeCapacity?: number;
    streamSpecification?: {
      streamEnabled: boolean;
      streamViewType: string;
    };
  }

  export type LifeCycleAction = 'create' | 'update' | 'destroy';

  export type QueryResponse = {
    Items: Item[];
    LastEvaluatedKey?: object;
    ConsumedCapacity?: ConsumedCapacity;
    Count: number;
    ScannedCount: number;
  };

  // Dynogels Model
  export interface Model {
    new (attrs: {[key: string]: any}): Item;

    get(hashKey: any, rangeKey: any, options: GetItemOptions, callback: DynogelsItemCallback): void;
    get(haskKey: any, options: GetItemOptions, callback: DynogelsItemCallback): void;
    get(hashKey: any, callback: DynogelsItemCallback): void;
    get(hashKey: any, rangeKey: any, callback: DynogelsItemCallback): void;
    create(item: any, options: CreateItemOptions, callback: DynogelsItemCallback): void;
    create(item: any, callback: DynogelsItemCallback): void;
    update(item: any, options: UpdateItemOptions, callback: DynogelsItemCallback): void;
    update(item: any, callback: DynogelsItemCallback): void;
    destroy(
      hashKey: any,
      rangeKey: any,
      options: DestroyItemOptions,
      callback: DynogelsItemCallback
    ): void;
    destroy(haskKey: any, options: DestroyItemOptions, callback: DynogelsItemCallback): void;
    destroy(hashKey: any, callback: DynogelsItemCallback): void;
    destroy(hashKey: any, rangeKey: any, callback: DynogelsItemCallback): void;
    destroy(item: any, options: DestroyItemOptions, callback: DynogelsItemCallback): void;
    destroy(item: any, callback: DynogelsItemCallback): void;
    query(hashKey: any): Query;
    scan(): Scan;
    parallelScan(totalSegments: number): Scan;
    getItems(
      items: string[] | Array<{[key: string]: string}>,
      callback: (err: Error, items: any[]) => void
    ): void;
    getItems(
      items: string[] | Array<{[key: string]: string}>,
      options: GetItemOptions,
      callback: (err: Error, items: any[]) => void
    ): void;
    batchGetItems(
      items: string[] | Array<{[key: string]: string}>,
      callback: (err: Error, items: any[]) => void
    ): void;
    batchGetItems(
      items: string[] | Array<{[key: string]: string}>,
      options: GetItemOptions,
      callback: (err: Error, items: any[]) => void
    ): void;
    createTable(
      options: CreateTablesOptions,
      callback: (err: Error, data: AWS.DynamoDB.CreateTableOutput) => void
    ): void;
    createTable(callback: (err: Error, data: AWS.DynamoDB.CreateTableOutput) => void): void;
    updateTable(
      throughput: Throughput,
      callback: (err: Error, data: AWS.DynamoDB.UpdateTableOutput) => void
    ): void;
    updateTable(callback: (err: Error, data: AWS.DynamoDB.UpdateTableOutput) => void): void;
    describeTable(callback: (err: Error, data: AWS.DynamoDB.DescribeTableOutput) => void): void;
    deleteTable(callback: (err: Error) => void): void;
    tableName(): string;

    after(action: LifeCycleAction, listner: (item: Item) => void): void;
    before(
      action: LifeCycleAction,
      listner: (data: any, next: (err: Error | null, data: any) => void) => void
    ): void;
    config(config: ModelConfig): {name: string};

    getAsync(hashKey: any, rangeKey: any, options: GetItemOptions): Promise<Item>;
    getAsync(haskKey: any, options: GetItemOptions): Promise<Item>;
    getAsync(hashKey: any): Promise<Item>;
    getAsync(hashKey: any, rangeKey: any): Promise<Item>;
    createAsync(item: any, options: CreateItemOptions): Promise<Item>;
    createAsync(item: any): Promise<Item>;
    updateAsync(item: any, options: UpdateItemOptions): Promise<Item>;
    updateAsync(item: any): Promise<Item>;
    destroyAsync(hashKey: any, rangeKey: any, options: DestroyItemOptions): Promise<Item>;
    destroyAsync(haskKey: any, options: DestroyItemOptions): Promise<Item>;
    destroyAsync(hashKey: any): Promise<Item>;
    destroyAsync(hashKey: any, rangeKey: any): Promise<Item>;
    destroyAsync(item: any, options: DestroyItemOptions): Promise<Item>;
    destroyAsync(item: any): Promise<Item>;
    parallelScanAsync(totalSegments: number): Scan;
    getItemsAsync(items: string[] | Array<{[key: string]: string}>): Promise<any[]>;
    getItemsAsync(
      items: string[] | Array<{[key: string]: string}>,
      options: GetItemOptions
    ): Promise<any[]>;
    batchGetItemsAsync(items: string[] | Array<{[key: string]: string}>): Promise<any[]>;
    batchGetItemsAsync(
      items: string[] | Array<{[key: string]: string}>,
      options: GetItemOptions
    ): Promise<any[]>;
    createTableAsync(options: CreateTablesOptions): Promise<AWS.DynamoDB.CreateTableOutput>;
    createTableAsync(): Promise<AWS.DynamoDB.CreateTableOutput>;
    updateTableAsync(throughput: Throughput): Promise<AWS.DynamoDB.UpdateTableOutput>;
    updateTableAsync(): Promise<AWS.DynamoDB.UpdateTableOutput>;
    describeTableAsync(): Promise<AWS.DynamoDB.DescribeTableOutput>;
    deleteTableAsync(): Promise<any>;
  }

  export type DynogelsItemCallback = (err: Error, data: Item) => void;

  export interface Throughput {
    readCapacity: number;
    writeCapacity: number;
  }

  export interface CreateItemOptions {
    expected?: {[key: string]: any};
    overwrite?: boolean;

    Expected?: AWS.DynamoDB.ExpectedAttributeMap;
    ReturnValues?: AWS.DynamoDB.ReturnValue;
    ReturnConsumedCapacity?: AWS.DynamoDB.ReturnConsumedCapacity;
    ReturnItemCollectionMetrics?: AWS.DynamoDB.ReturnItemCollectionMetrics;
    ConditionalOperator?: AWS.DynamoDB.ConditionalOperator;
    ConditionExpression?: AWS.DynamoDB.ConditionExpression;
    ExpressionAttributeNames?: AWS.DynamoDB.ExpressionAttributeNameMap;
    ExpressionAttributeValues?: {[key: string]: any};
  }

  export interface UpdateItemOptions {
    expected?: {[key: string]: any};

    AttributeUpdates?: AWS.DynamoDB.AttributeUpdates;
    Expected?: AWS.DynamoDB.ExpectedAttributeMap;
    ConditionalOperator?: AWS.DynamoDB.ConditionalOperator;
    ReturnValues?: AWS.DynamoDB.ReturnValue;
    ReturnConsumedCapacity?: AWS.DynamoDB.ReturnConsumedCapacity;
    ReturnItemCollectionMetrics?: AWS.DynamoDB.ReturnItemCollectionMetrics;
    UpdateExpression?: AWS.DynamoDB.UpdateExpression;
    ConditionExpression?: AWS.DynamoDB.ConditionExpression;
    ExpressionAttributeNames?: AWS.DynamoDB.ExpressionAttributeNameMap;
    ExpressionAttributeValues?: {[key: string]: any};
  }

  export interface DestroyItemOptions {
    Expected?: AWS.DynamoDB.ExpectedAttributeMap;
    ConditionalOperator?: AWS.DynamoDB.ConditionalOperator;
    ReturnValues?: AWS.DynamoDB.ReturnValue;
    ReturnConsumedCapacity?: AWS.DynamoDB.ReturnConsumedCapacity;
    ReturnItemCollectionMetrics?: AWS.DynamoDB.ReturnItemCollectionMetrics;
    ConditionExpression?: AWS.DynamoDB.ConditionExpression;
    ExpressionAttributeNames?: AWS.DynamoDB.ExpressionAttributeNameMap;
    ExpressionAttributeValues?: {[key: string]: any};
  }

  export interface GetItemOptions {
    AttributesToGet?: AWS.DynamoDB.AttributeNameList;
    ConsistentRead?: AWS.DynamoDB.ConsistentRead;
    ReturnConsumedCapacity?: AWS.DynamoDB.ReturnConsumedCapacity;
    ProjectionExpression?: AWS.DynamoDB.ProjectionExpression;
    ExpressionAttributeNames?: AWS.DynamoDB.ExpressionAttributeNameMap;
  }

  export interface ModelConfig {
    tableName?: string;
    docClient?: any;
    dynamodb?: AWS.DynamoDB;
  }

  // Dynogels Item
  export interface Item {
    get(): {[key: string]: any};
    get(key: string): any;
    set(params: {}): Item;
    save(callback?: DynogelsItemCallback): void;
    update(options: UpdateItemOptions, callback?: DynogelsItemCallback): void;
    update(callback?: DynogelsItemCallback): void;
    destroy(options: DestroyItemOptions, callback?: DynogelsItemCallback): void;
    destroy(callback?: DynogelsItemCallback): void;
    toJSON(): any;
    toPlainObject(): any;
  }

  export interface BaseChain<T> {
    equals(value: any): T;
    eq(value: any): T;
    lte(value: any): T;
    lt(value: any): T;
    gte(value: any): T;
    gt(value: any): T;
    null(): T;
    exists(): T;
    beginsWith(value: any): T;
    between(value1: any, value2: any): T;
  }

  export interface ExtendedChain<T> extends BaseChain<T> {
    contains(value: any): T;
    notContains(value: any): T;
    in(values: any[]): T;
    ne(value: any): T;
  }

  export type QueryWhereChain = BaseChain<Query>;
  export type QueryFilterChain = ExtendedChain<Query>;

  // Dynogels Query
  export interface Query {
    limit(number: number): Query;
    filterExpression(expression: any): Query;
    expressionAttributeNames(data: any): Query;
    expressionAttributeValues(data: any): Query;
    projectionExpression(data: any): Query;
    usingIndex(name: string): Query;
    consistentRead(read: boolean): Query;
    addKeyCondition(condition: any): Query;
    addFilterCondition(condition: any): Query;
    startKey(hashKey: any, rangeKey?: any): Query;
    attributes(attrs: any): Query;
    ascending(): Query;
    descending(): Query;
    select(value: any): Query;
    returnConsumedCapacity(value: any): Query;
    loadAll(): Query;
    where(keyName: string): QueryWhereChain;
    filter(keyName: string): QueryFilterChain;
    exec(): stream.Readable;
    exec(callback: (err: Error, data: QueryResponse) => void): void;
    execAsync(): Promise<QueryResponse>;
  }

  export interface ScanWhereChain extends ExtendedChain<Scan> {
    notNull(): Scan;
  }

  // Dynogels Scan
  export interface Scan {
    limit(number: number): Scan;
    addFilterCondition(condition: any): Scan;
    startKey(hashKey: any, rangeKey?: any): Scan;
    attributes(attrs: any): Scan;
    select(value: any): Scan;
    returnConsumedCapacity(): Scan;
    segments(segment: any, totalSegments: number): Scan;
    where(keyName: string): ScanWhereChain;
    filterExpression(expression: any): Scan;
    expressionAttributeNames(data: any): Scan;
    expressionAttributeValues(data: any): Scan;
    projectionExpression(data: any): Scan;
    exec(): stream.Readable;
    exec(callback: (err: Error, data: QueryResponse) => void): void;
    loadAll(): Scan;
    execAsync(): Promise<QueryResponse>;
  }

  export type tableResolve = () => string;

  export interface SchemaType {
    [key: string]: joi.AnySchema | SchemaType;
  }

  export interface ModelConfiguration {
    hashKey: string;
    rangeKey?: string;
    timestamps?: boolean;
    createdAt?: boolean;
    updatedAt?: string;
    schema?: SchemaType;
    validation?: joi.ValidationOptions;
    tableName?: string | tableResolve;
    indexes?: any[];
    log?: Log;
  }

  export interface Document {
    [key: string]: any;
  }

  export interface DocumentCollection {
    Items: Document[];
    Count: number;
    ScannedCount: number;
  }
}
