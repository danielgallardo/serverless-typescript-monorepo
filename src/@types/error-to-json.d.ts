declare module 'error-to-json' {
  export type JSONError = {
    name: string;
    message: string;
    stack?: string[];
    [key: string]: any;
  };

  function errorToJSON(error: Error): JSONError;

  export default errorToJSON;
}
