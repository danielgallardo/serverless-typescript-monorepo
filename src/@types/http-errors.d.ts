declare module 'http-errors' {
  interface IHttpError extends Error {
    status: number;
    statusCode: number;
    expose: boolean;
    headers?: {
      [key: string]: string;
    };
    [key: string]: any;
  }

  type HttpErrorConstructor = new (msg?: string) => IHttpError;

  type CreateHttpError = (
    ...args: Array<Error | string | number | {[key: string]: any}>
  ) => IHttpError;

  type NamedConstructors = {
    [code: string]: HttpErrorConstructor;
    HttpError: HttpErrorConstructor;
    BadRequest: HttpErrorConstructor;
    Unauthorized: HttpErrorConstructor;
    PaymentRequired: HttpErrorConstructor;
    Forbidden: HttpErrorConstructor;
    NotFound: HttpErrorConstructor;
    MethodNotAllowed: HttpErrorConstructor;
    NotAcceptable: HttpErrorConstructor;
    ProxyAuthenticationRequired: HttpErrorConstructor;
    RequestTimeout: HttpErrorConstructor;
    Conflict: HttpErrorConstructor;
    Gone: HttpErrorConstructor;
    LengthRequired: HttpErrorConstructor;
    PreconditionFailed: HttpErrorConstructor;
    PayloadTooLarge: HttpErrorConstructor;
    URITooLong: HttpErrorConstructor;
    UnsupportedMediaType: HttpErrorConstructor;
    RangeNotSatisfiable: HttpErrorConstructor;
    ExpectationFailed: HttpErrorConstructor;
    ImATeapot: HttpErrorConstructor;
    MisdirectedRequest: HttpErrorConstructor;
    UnprocessableEntity: HttpErrorConstructor;
    Locked: HttpErrorConstructor;
    FailedDependency: HttpErrorConstructor;
    UnorderedCollection: HttpErrorConstructor;
    UpgradeRequired: HttpErrorConstructor;
    PreconditionRequired: HttpErrorConstructor;
    TooManyRequests: HttpErrorConstructor;
    RequestHeaderFieldsTooLarge: HttpErrorConstructor;
    UnavailableForLegalReasons: HttpErrorConstructor;
    InternalServerError: HttpErrorConstructor;
    NotImplemented: HttpErrorConstructor;
    BadGateway: HttpErrorConstructor;
    ServiceUnavailable: HttpErrorConstructor;
    GatewayTimeout: HttpErrorConstructor;
    HTTPVersionNotSupported: HttpErrorConstructor;
    VariantAlsoNegotiates: HttpErrorConstructor;
    InsufficientStorage: HttpErrorConstructor;
    LoopDetected: HttpErrorConstructor;
    BandwidthLimitExceeded: HttpErrorConstructor;
    NotExtended: HttpErrorConstructor;
    NetworkAuthenticationRequire: HttpErrorConstructor;
  };

  const createHttpError: CreateHttpError & NamedConstructors;
  export = createHttpError;
}
