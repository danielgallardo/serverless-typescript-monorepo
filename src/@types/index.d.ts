import {APIGatewayEventRequestContext} from 'aws-lambda';

export interface NormalizedEvent {
  headers: {
    [name: string]: string;
  };
  httpMethod: string;
  isBase64Encoded: boolean;
  path: string;
  pathParameters: {[name: string]: string} | {};
  queryStringParameters: {[name: string]: string} | {};
  stageVariables: {[name: string]: string} | null;
  requestContext: APIGatewayEventRequestContext;
  resource: string;
  body: any;
}
