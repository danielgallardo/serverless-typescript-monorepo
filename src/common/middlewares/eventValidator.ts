import {IHandlerLambda, IMiddyNextFunction} from 'middy';
import {SchemaLike, ValidationOptions} from 'joi';
import {attempt} from '../validation';

const defaultOptions = {
  stripUnknown: true
};

export const eventValidator = (schema: SchemaLike, options?: ValidationOptions) => ({
  before(handler: IHandlerLambda, next: IMiddyNextFunction) {
    const _options = Object.assign({}, defaultOptions, options);
    const event = attempt(handler.event, schema, undefined, _options);
    Object.assign(handler.event, event);
    next();
  }
});
