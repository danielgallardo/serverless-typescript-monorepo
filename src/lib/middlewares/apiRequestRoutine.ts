import {APIGatewayProxyEvent} from 'aws-lambda';
// @ts-ignore
import contentTypeLib from 'content-type';
import createError, {HttpError, ServiceUnavailable, UnprocessableEntity} from 'http-errors';
import {ValidationError} from 'joi';
import {IHandlerLambda, IMiddyNextFunction} from 'middy';
import {JoiError} from '../errors';
import {errorToJSON} from '../logger';

const DEFAULT_HEADERS = {
  'Access-Control-Allow-Origin': '*', // Required for CORS support to work
  'Access-Control-Allow-Credentials': true // Required for cookies, authorization headers with HTTPS
};

/**
 * Checks if body is a valid JSON and converts it to object
 * @param event
 */
export const parseJSONBody = (event: APIGatewayProxyEvent): void => {
  if (!event.headers || !event.body) return;
  const contentType = event.headers['Content-Type'] || event.headers['content-type'];
  if (contentType) {
    const {type} = contentTypeLib.parse(contentType);
    if (type === 'application/json') {
      try {
        event.body = JSON.parse(event.body);
      } catch (err) {
        throw new UnprocessableEntity(
          'Content type defined as JSON but an invalid JSON was provided'
        );
      }
    }
  }
};

/**
 * Sets queryStringParameters and pathParameters to {} by default
 * @param event
 */
export const setDefaultParams = (event: APIGatewayProxyEvent): void => {
  if (event.hasOwnProperty('httpMethod')) {
    event.queryStringParameters = event.queryStringParameters || {};
    event.pathParameters = event.pathParameters || {};
  }
};

interface IResponse {
  statusCode?: number;
  [key: string]: any;
}

/**
 * Builds a valid ApiGateway response based on returned value
 * redirects if response contains redirectUrl
 * @param response
 */
export const buildJSONResponse = (response?: IResponse): IResponse => {
  // return No Content response if there is no content
  if (!response) {
    return {
      statusCode: 204,
      headers: DEFAULT_HEADERS
    };
  }

  // return response directly if it contains statusCode
  if (response.statusCode) {
    return {
      ...response,
      headers: {
        ...DEFAULT_HEADERS,
        ...response.headers
      }
    };
  }

  // return redirect response if redirectUrl is defined
  if (response.redirectUrl) {
    return {
      statusCode: 302,
      headers: {
        ...DEFAULT_HEADERS,
        Location: response.redirectUrl
      }
    };
  }

  return {
    statusCode: 200,
    body: JSON.stringify(response),
    headers: DEFAULT_HEADERS
  };
};

/**
 * Creates a public ServiceUnavailable error
 * provides original error details if env "DEBUG" is set to true
 * @param error
 */
const getServiceUnavailableError = (error: Error) => {
  const httpError = new ServiceUnavailable();
  if (process.env.DEBUG === 'true') {
    httpError.details = errorToJSON(error);
  }
  return httpError;
};

/**
 * Gets public error based on error type
 * @param error
 */
const getPublicError = (error: Error) => {
  if (error instanceof HttpError) return error;
  if (error instanceof JoiError) {
    if (error.internal) return getServiceUnavailableError(error);
    else return createError(422, error.message, {details: error.details});
  }
  if (error.name === 'ValidationError') {
    const validationError = new JoiError(error as ValidationError);
    return createError(422, validationError.message, {details: validationError.details});
  }
  return getServiceUnavailableError(error);
};

/**
 * Builds a valid ApiGateway error response based on error
 * @param error
 */
export const buildErrorResponse = (error: Error): IResponse => {
  const httpError = getPublicError(error);
  return {
    statusCode: httpError.statusCode,
    headers: DEFAULT_HEADERS,
    body: JSON.stringify({
      status: httpError.name,
      statusCode: httpError.statusCode,
      message: httpError.message,
      details: httpError.details
    })
  };
};

/**
 * A middleware that includes useful logic for ApiGateway handlers
 */
export const apiRequestRoutine = () => ({
  before(handler: IHandlerLambda, next: IMiddyNextFunction) {
    parseJSONBody(handler.event);
    setDefaultParams(handler.event);
    return next();
  },

  after(handler: IHandlerLambda, next: IMiddyNextFunction) {
    handler.response = buildJSONResponse(handler.response);
    return next();
  },

  onError(handler: IHandlerLambda, next: IMiddyNextFunction) {
    handler.response = buildErrorResponse(handler.error);
    return next();
  }
});
