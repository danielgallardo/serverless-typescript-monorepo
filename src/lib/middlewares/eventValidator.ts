import {SchemaLike, ValidationOptions} from 'joi';
import {IHandlerLambda, IMiddyNextFunction} from 'middy';
import {attempt} from '../validation';

const defaultOptions = {
  stripUnknown: true
};

/**
 * Validates event based on Joi schema
 * @param schema
 * @param options
 */
export const eventValidator = (schema: SchemaLike, options?: ValidationOptions) => ({
  before(handler: IHandlerLambda, next: IMiddyNextFunction) {
    const _options = Object.assign({}, defaultOptions, options);
    const event = attempt(handler.event, schema, undefined, _options);
    Object.assign(handler.event, event);
    next();
  }
});
