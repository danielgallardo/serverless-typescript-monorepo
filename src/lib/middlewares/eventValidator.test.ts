import clone from 'clone';
import Joi from 'joi';
import middy from 'middy';
import {INormalizedEvent} from '../../@types';
import {normalizedEvent} from '../fixtures/events';
import {apiRequestRoutine} from './apiRequestRoutine';
import {eventValidator} from './eventValidator';

describe('eventValidator', () => {
  let demoEvent: INormalizedEvent;
  beforeEach(() => {
    demoEvent = clone(normalizedEvent);
  });

  it('should pass validation with correct schema', () => {
    demoEvent.body = {foo: 'bar'};
    demoEvent.pathParameters = {path: 'parameter'};
    demoEvent.queryStringParameters = {query: 'parameter'};

    const schema = {
      body: Joi.object().keys({
        foo: Joi.string().required()
      }),
      pathParameters: Joi.object().keys({
        path: Joi.string().required()
      }),
      queryStringParameters: Joi.object().keys({
        query: Joi.string().required()
      })
    };

    const handler = middy(async () => {
      return {foo: 'bar'};
    });
    handler.use(eventValidator(schema));

    // @ts-ignore
    handler(demoEvent, {}, (error, response) => {
      expect(error).toBeNull();
      expect(response).toMatchSnapshot();
    });
  });

  it('should strip unknown params by default', () => {
    demoEvent.body = {foo: 'bar', extra: 'param'};

    const schema = {
      body: Joi.object().keys({
        foo: Joi.string().required()
      })
    };

    const handler = middy(async (event: INormalizedEvent) => {
      expect(event.body).toEqual({foo: 'bar'});
      return {foo: 'bar'};
    });
    handler.use(eventValidator(schema));

    // @ts-ignore
    handler(demoEvent, {}, (error, response) => {
      expect(error).toBeNull();
      expect(response).toMatchSnapshot();
    });
  });

  it('should be compatable with apiRequestRoutine', () => {
    demoEvent.body = JSON.stringify({foo: 'bar'});

    const schema = {
      body: {
        foo: Joi.string().required()
      }
    };

    const handler = middy(async () => {
      return {foo: 'bar'};
    });
    handler.use(apiRequestRoutine()).use(eventValidator(schema));

    // @ts-ignore
    handler(demoEvent, {}, (error, response) => {
      expect(error).toBeNull();
      expect(response).toMatchSnapshot();
    });
  });

  it('should respond with http error with apiRequestRoutine when schema is invalid', () => {
    demoEvent.body = JSON.stringify({foo: 'bar'});

    const schema = {
      body: Joi.object().keys({
        bar: Joi.string().required()
      })
    };

    const handler = middy(async () => {
      return {foo: 'bar'};
    });
    handler.use(apiRequestRoutine()).use(eventValidator(schema));

    // @ts-ignore
    handler(demoEvent, {}, (error, response) => {
      expect(error).toBeNull();
      expect(response).toMatchSnapshot();
    });
  });
});
