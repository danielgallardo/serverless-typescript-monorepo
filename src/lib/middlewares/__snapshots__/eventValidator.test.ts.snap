// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eventValidator should be compatable with apiRequestRoutine 1`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;

exports[`eventValidator should pass validation with correct schema 1`] = `
Object {
  "foo": "bar",
}
`;

exports[`eventValidator should respond with http error with apiRequestRoutine when schema is invalid 1`] = `
Object {
  "body": "{\\"status\\":\\"UnprocessableEntityError\\",\\"statusCode\\":422,\\"message\\":\\"\\\\\\"bar\\\\\\" is required\\",\\"details\\":[{\\"message\\":\\"\\\\\\"bar\\\\\\" is required\\",\\"path\\":[\\"body\\",\\"bar\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"bar\\",\\"label\\":\\"bar\\"}}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 422,
}
`;

exports[`eventValidator should strip unknown params by default 1`] = `
Object {
  "foo": "bar",
}
`;
